recipes/automated-testing.md

```diff
@@ -0,0 +1,239 @@
+follow the business logic instead of programming logic. Avoid writing parts of
+the function implementation in the actual test assertion. This will lead to tight
+coupling of tests with internal implementation and the tests will have to be fixed
+each time the logic changes.
+
+---
+
+Writing quality tests is hard and it's easy to fall into common pitfalls of testing
+that the database update function actually updates the database. Start off simple
+and as the application grows in complexity, it will be easier to determine what
+should be tested more thoroughly. It is perfectly fine to have a small test suite
+that covers the critical code and the essentials. Small suites will run faster
+which means they will be run more often.
+
+## Types of Automated Tests
+
+There are different approaches to testing, and depending on boundaries of the
+test, we can split them into following categories:
+
+- **Unit Tests**
+- **Integration Tests**
+- **API Tests**
+- **E2E Tests**
+- **Load/Performance Tests**
+- **Visual Tests**
+
+*Note that some people can call these tests by different names, but for Studion
+internal purposes, this should be considered the naming convention.*
+
+### Unit Tests
+
+These are the most isolated tests that we can write. They should take a specific
+function/service/helper/module and test its functionality. Unit tests will
+usually require mocked data, but since we're testing the case when specific
+input produces specific output, the mocked data set should be minimal.
+
+Unit testing is recommended for functions that contain a lot of logic and/or branching.
+It is convenient to test a specific function at the lowest level so if the logic
+changes, we can make minimal changes to the test suite and/or mocked data.
+
+#### When to use
+- Test a unit that implements the business logic, that's isolated from side effects such as database interaction or HTTP request processing
+- Test function or class method with multiple input-output permutations
+
+#### When **not** to use
+- To test unit that integrates different application layers, such as persistence layer (database) or HTTP layer (see "Integration Tests") or performs disk I/O or communicates with external system
+
+#### Best practices
+- Unit tests should execute fast (<50ms)
+- Use mocks and stubs through dependency injection (method or constructor injection)
+
+#### Antipatterns
+- Mocking infrastructure parts such as database I/O - instead, revert the control by using the `AppService`, `Command` or `Query` to integrate unit implementing business logic with the infrastructure layer of the application
+- Monkey-patching dependencies used by the unit - instead, pass the dependencies through the constructor or method, so that you can pass the mocks or stubs in the test
+
+
+### Integration Tests
+
+With these tests, we test the application API endpoints and assert that they are
+actually working as expected.
```

Comment:
Currently, the description of integration and API tests is confusing and incorrect. Integration tests are meant to test the application's components, while API tests are intended to test the public API, i.e. contract. Mentioning API in the integration tests is misleading.

One other thing I would mention in the integration tests is the level at which we do the testing. Since all our projects to my knowledge use MVC, i.e. service/controller nomenclature, stating to do the integration tests on the service level is beneficial and saves time while things in the controller get tested by the API tests, at least that is my experience.
@dbettini You have a lot of experience with integration testing in general, what is your opinion on this and integration testing in general, how does this look to you, would you add something, or have some advice?

Reply:
From where I see it, the easiest entry point for integration tests is API. And since controllers are just wrappers for services, we can easily test the API with integration tests (check for authentication and authorization and make sure the correct HTTP codes are returned) and also add tests for service logic, for example check the DB and make sure calling the API actually made the change.
I might be wrong, so I'd like to hear what other people think. Also, I think we should schedule a call next week to talk about this and resolve any questions.

Reply:
This is how I see things, integration tests are used to test a group of components as a whole, and API tests are a subset of integration tests that specifically focus on testing a public contract used by other applications for communication. In some cases when the backend is a separate application, API tests can even be considered as E2E tests as they test the application from end to end.
Having said that, the equalization of integration tests and API tests is limiting and narrows the scope of integration we want to test as some integrations between components are private, e.g. service called by the corn job.
Testing authentication, authorization, status codes, and input validation should all be a part of the API test as it describes the contract, e.g. if not authenticated return 401, if the input is not valid JSON return 400, if authenticated and input is valid JSON return 200, etc.

Reply:
As you've both mentioned, API tests can be considered a subset of integration tests, since they test the integration of multiple components. And, in some cases (e.g. NestJS docs), they are also called E2E tests (I'm not a fan of that nomenclature, but I get where it's coming from).

As for the specific entry point for the tests, I don't think there's only one correct answer - there are pros and cons to both. Examples:
1. Use case can be invoked by both API and a console command - In this case, it might make sense to write integration tests on the use case (service or controller, since it should be thin) level. The problem is that e.g. auth can't be tested by invoking use cases directly, so either another layer of tests should still be added to cover different protocols, or the tests should be duplicated instead of invoking the use cases directly. The former is usually picked here, since the duplication gets out of hand pretty quickly. In that case, middlewares/guards/filters applied by the framework on top of the use cases must still be tested in the upper layer (e.g. API).
2. Use case can be invoked by the API only, and it's highly unlikely that it will change - This is the most common situation I've bumped into. In this case, the previous approach could definitely be used - however, the most straightforward way is to just use API tests and eliminate one level entirely. One example I can think of where this was useful was using too small of a gRPC message integer size (which resulted in integer overflowing into negative value), which was caught in tests - this would not be possible if the tests didn't go through gRPC, mimicking the actual app behavior as close as possible. 

As for business logic heavy use cases - I've actually started moving it from services into models, which are then unit tested. This is because doing integration tests (regardless of their entry) can easily get out of hand as the number of use cases grow, raising the number of possible permutations within them exponentially. So, the approach that has been working for me so far is extracting the business logic into unit testable models, and doing API tests for integration. But to do that correctly, it's crucial that:
- business logic models are isolated from the storage infrastructure
- controllers are thin and don't bring in additional logic compared to the services
- integration tests use a full testing infrastructure (regardless of their entry point), allowing testing of CRUD actions and third party service calls

 But like I said, that is just what worked in these particular cases - being pragmatic here is crucial. Just like a dev needs to decide how many integration permutations are enough and leave the rest to unit tests, they should also decide which approach to take regarding the entry point. That is why I'd rather mention these options here with their pros and cons, instead of pointing only to API tests in the integration testing section (although I think most of the projects will benefit from that approach the most).

Reply:
@dbettini What would you change/add to the integration and API testing sections of this guide to align them with your comment, which I find useful, concrete, and on point?

Reply:
@ikovac ^

Reply:
Done ✅